resources/chain.rb

```diff
@@ -1,47 +1,71 @@
-#
-# Author:: Ben Hughes <bmhughes@bmhughes.co.uk>
-# Cookbook:: iptables
-# Resource:: chain
-#
-# Copyright:: 2019, Ben Hughes
-# Copyright:: 2017-2019, Chef Software, Inc.
-#
-# Licensed under the Apache License, Version 2.0 (the "License");
-# you may not use this file except in compliance with the License.
-# You may obtain a copy of the License at
-#
-#     http://www.apache.org/licenses/LICENSE-2.0
-#
-# Unless required by applicable law or agreed to in writing, software
-# distributed under the License is distributed on an "AS IS" BASIS,
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License.
-
-property :source, String, default: 'iptables.erb'
-property :cookbook, String, default: 'iptables'
-property :config_file, String, default: lazy { node['iptables']['persisted_rules_iptables'] }
-property :table, String, equal_to: %w(filter mangle nat raw security), default: 'filter'
-property :chain, [String, Array, Hash]
-property :filemode, [String, Integer], default: '0644'
+include Iptables::Cookbook::Helpers
+
+property :table, [Symbol, String],
+          equal_to: [:filter, :mangle, :nat, :raw, :security, 'filter', 'mangle', 'nat', 'raw', 'security'],
+          default: :filter,
+          description: 'The table the chain should exist on'
+
+property :chain, [Symbol, String],
+          description: 'The name of the Chain'
+
+property :value, String,
+          default: 'ACCEPT [0:0]',
+          description: 'The default action and the Packets : Bytes count'
+
+property :ip_version, Symbol,
+          equal_to: %i(ipv4 ipv6),
+          default: :ipv4,
+          description: 'The IP version, 4 or 6'
+
+property :source_template, String,
+          default: 'iptables.erb',
+          description: 'Source template to use to create the rules'
+
+property :cookbook, String,
+          default: 'iptables',
+          description: 'Source cookbook to find the template in'
+
+property :sensitive, [true, false],
+          default: false,
+          description: 'mark the resource as senstive'
+
+property :config_file, String,
+          default: lazy { default_iptables_rules_file(ip_version) },
+          description: 'The full path to find the rules on disk'
 
 action :create do
-  Chef::Resource::Template.send(:include, Iptables::ChainHelpers)
+  # We are using the accumalator pattern here
+  # This is as we are managing a single config file but using multiple
+  # resouces to allow a cleaner api for the end user
+  # Note, this will only ever go as a file on disk at the end of a chef run
+  if new_resource.table.class == 'String'
```

Comment:
It's probably better to just use a proc to set the string to a symbol and then validate on that.